README
PHP extension for ffmpeg
========================

Description
-----------
ffmpeg-php is an extension for PHP that adds an easy to use, object-oriented API for accessing and retrieving information from video and audio files. It has methods for returning frames from movie files as images that can be manipulated using PHP's image functions. This works well for automatically creating thumbnail images from movies. ffmpeg-php is also useful for reporting the duration and bitrate of audio files (mp3, wma...). ffmpeg-php can access many of the video formats supported by ffmpeg (mov, avi, mpg, wmv...).

Originally hosted at http://ffmpeg-php.sourceforge.net, the project seems to be abandoned right now.
So here is a fork of it with minor modifications required to make it work with recent PHP versions.

This is a working fork of the tony2001/ffmpeg-php project that appears to be abandoned as well. The objective of this project fork at this point is fixing depracated constants against the latest build of ffmpeg.


ChangeLog
ChangeLog for ffmpeg-php (ffmpeg-php.sourceforge.net)

version 0.7.0 (1/24/2018)
    Bug Fixes
    * Backport fixes

version 0.6.3 (3/30/2009)
    Bug Fixes
    * Fixed compatibility with ffmpeg 0.5.0 release. ffmpeg-php is only
    supporting offical releases from here on.

version 0.6.2 (3/25/2009)
    Bug Fixes
    * Fixed building on php4
    * Fixed compatibility with ffmpeg 0.5.0 release 

version 0.6.0 (10/15/2008)
    New Features:
    * Support for ffmpeg swscale API

    API Changes:
    * Removed obsolete resize() and crop() functions. These operations are
      more flexibly done using the gd commands.
    * Removed animated gif class since it was way out of date and the 
      quality of animated gifs created by ffmpeg sucks anyway

    Bug Fixes:
    * Added check to ensure that gd loads before ffmpeg-php if both are
      compiled as extensions.
    * Replaced large mp3 test file with smaller sample to avoid copyright
      issues and reduce package size.
    * Convert from deprecated img_* functions to swscale.
    * Builds against php4 correctly again.

version 0.5.3 (05/30/2008)
    New Features:
    * Added 'ffmpeg.show_warnings' ini flag to control whether ffmpeg 
      warnings are converted to php warnings.
    * Applied patch to add getAudioStreamId and getVideoStreamId functions 

version 0.5.2.1 (04/03/2008)

    Bug Fixes:
    * Minor build fixes and cleanups

version 0.5.2 (04/02/2008)

    Bug Fixes:
    * ffmpeg-php should compile on macos-x now
    * Failure to find the proper codec is not a fatal error anymore
    * Updated build script to deal with new ffmpeg include file layout.

    
    New Features:
    * Added gd image bounds checking.
    * Added config flag to skip the compile-time gd check.
    * ffmpeg's logs are mapped to php's warnings and notices (no more stderr pollution)

version 0.5.1 (06/11/2007)

    Bug Fixes:
    * Fixed some of the tests (Thanks to Raphael Geissert for the patch)
    * Improved build process (Thanks to Raphael Geissert for the suggestions)
    * Fixed the test script so it won't generate warnings when video/audio
      stream is missing.

    New Features:
    * Added $movie->hasVideo() function to test for video stream.
    * Copyright updated for better compatibilty with the PHP license.

version 0.5.0 (09/29/2006)

    Bug Fixes:
    * Fixed problem with params to av_open_input_file not getting initialized
      correctly. (Thanks to SilverCode for the fix)

version 0.4.9 (09/04/2006)

    New Features:
    * Better error reporting when opening a movie fails
    * INSTALL doc clarifications

    Bug Fixes:
    * Fixed compilation error in ffmpeg_animated_gif.c when building against 
      old ffmpeg


version 0.4.8 (08/04/2006)
    
    New Features:
    * Added new functions for audio streams: getAudioSampleRate(), 
      getAudioBitRate() (Martin Prangl)
    * Changed getBitRate function to like getAudioSampleRate() and
      getAudioBitRate(). Now returns bits per second instead of
      calculating kb/s. (To get kb/s just do movie->getBitRate() / 1000
      in php.) (tkirby)

    Bug Fixes:
    * Fix possible error in hasAudio calculation if audio stream is the
      first stream. (tkirby)

version 0.4.7 (06/08/2006)

    Bug Fixes:
    * Fixed framerate calculation to be compatible with ffmpeg cvs 6/8/06 
      (tkirby)

version 0.4.6 (08/30/2005)

    Bug Fixes:
    * Fixed compatiblity with ffmpeg cvs after 7/30/05 or so (tkirby)

version 0.4.5 (07/12/2005)
    Bug Fixes:
    * Minor build fixes for older ffmpeg versions (tkirby)

version 0.4.4 (06/24/2005)
    New Features:
    * Added ability to create animated gifs via new ffmpeg_animated_gif
      class (tkirby)
    * Fixed ffmpeg_frame constructor to allow creating ffmpeg_frames from
      gd images. This allows to make animated gifs directly from a series
      of gd images. I believe this is currently the only way to make 
      animated gifs using PHP since the bundled version of gd does not
      support animated gif creation (tkirby)
    * Added flag to php.ini to allow server admins to disable persistent
      movies since injudicious use of persistent movies in scripts can eat 
      up server resources quickly (tkirby)
    * Added frame->isKeyFrame() and movie->getNextKeyFrame() methods (tkirby)
    * Now displays ffmpeg-php ini values in php_info() table (tkirby)

    Bug Fixes:
    * Fixed parsing of 'persistent' boolean flag in ffmpeg_movie constructor
    * Fixed segfault when a script tries to read more frames than exist in
      a movie. (tkirby)
    * Patch to use native time bases. This allows building against libavcodec 
      build 4754+ (Steve Brown)
    * Patch to corrctly init gd before attempting to fetch gd image resources.
      This fixes the "togdimage() supplied resource is not a valid Image 
      resource" error that a few people have reported (Steve Brown)

version 0.4.3 (03/12/2005)
    Bug Fixes:
    * Added ability to open movies as persistent resources. This greatly
      increases speed of repeatedly accessing the same movie from separate
      php scripts. This is still somewhat experimental so don't use it
      unless you have a specific need (tkirby)
    * Increased strictness of compiler warnings to catch issues like
      undefined symbol errors. (tkirby)
    * Track frame number within ffmpeg-php since the frame_number variable
      in ffmpeg's decoder context seems to give inaccurate results on some 
      3GP movies. (tkirby)
    * Fixed bug #1099229 - Stride problems if ffmpeg_frame is resized upwards
      (tkirby)
    * Functions now use safe_emalloc() to avoid possible integer overflows 
      (tkirby)
    * Changed lrint() to a macro to allow compiling on Windows (tkirby)

version 0.4.2.1 (02/15/2005)
    Bug Fixes:
    * Fixed error handling when ffmpeg_movie object creation fails (tkirby)
    * Fixed undefined symbol error in ffmpeg_output_movie (tkirby)


php_ffmpeg.h
/*
   This file is part of ffmpeg-php

   Copyright (C) 2004-2008 Todd Kirby (ffmpeg.php AT gmail.com)

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

   In addition, as a special exception, the copyright holders of ffmpeg-php
   give you permission to combine ffmpeg-php with code included in the
   standard release of PHP under the PHP license (or modified versions of
   such code, with unchanged license). You may copy and distribute such a
   system following the terms of the GNU GPL for ffmpeg-php and the licenses
   of the other code concerned, provided that you include the source code of
   that other code when and as the GNU GPL requires distribution of source code.

   You must obey the GNU General Public License in all respects for all of the
   code used other than standard release of PHP. If you modify this file, you
   may extend this exception to your version of the file, but you are not
   obligated to do so. If you do not wish to do so, delete this exception
   statement from your version.

 */

#ifndef PHP_FFMPEG_H
#define PHP_FFMPEG_H

#include "php_version.h"


/* 
 * PHP-4 doesn't have the METHOD, ME and MALIAS macros so map them back
 * to the function macros for PHP-4
 */

#if PHP_MAJOR_VERSION <= 4 // PHP4; use FUNCTION macros
#define FFMPEG_PHP_CONSTRUCTOR(a,b) PHP_FUNCTION(a)
#define FFMPEG_PHP_METHOD(a,b) PHP_FUNCTION(b)
#define FFMPEG_PHP_ME(a,b,c,d) PHP_FE(a,c)
#define FFMPEG_PHP_MALIAS(a,b,c,d,e) PHP_FALIAS(b,c,d)
#define FFMPEG_PHP_END_METHODS {NULL, NULL, NULL}
#else /* PHP5; Use METHOD macros */
#define FFMPEG_PHP_CONSTRUCTOR(a,b) PHP_METHOD(a,b)
#define FFMPEG_PHP_METHOD(a,b) PHP_METHOD(a,b)
#define FFMPEG_PHP_ME(a,b,c,d) PHP_ME(a,b,c,d)
#define FFMPEG_PHP_MALIAS(a,b,c,d,e) PHP_MALIAS(a,b,c,d,e)
#define FFMPEG_PHP_END_METHODS {NULL, NULL, NULL, 0, 0}
#endif

/*
 * NEW fix for `undefined symbol: avcodec_alloc_frame in Unknown on line 0`
 */
#define avcodec_alloc_frame av_frame_alloc


#define SAFE_STRING(s) ((s)?(s):"")

#ifndef safe_emalloc
    # define safe_emalloc(a,b,c) emalloc((a)*(b)+(c))
#endif

extern zend_module_entry ffmpeg_module_entry;
#define phpext_ffmpeg_ptr &ffmpeg_module_entry

#ifdef PHP_WIN32
#define PHP_FFMPEG_API __declspec(dllexport)
#else
#define PHP_FFMPEG_API
#endif

#ifdef ZTS
#include "TSRM.h"
#endif

PHP_MINIT_FUNCTION(ffmpeg);
PHP_MSHUTDOWN_FUNCTION(ffmpeg);
PHP_RINIT_FUNCTION(ffmpeg);
PHP_RSHUTDOWN_FUNCTION(ffmpeg);
PHP_MINFO_FUNCTION(ffmpeg);


#ifdef ZTS
#define FFMPEG_G(v) TSRMG(ffmpeg_globals_id, zend_ffmpeg_globals *, v)
#else
#define FFMPEG_G(v) (ffmpeg_globals.v)
#endif

#endif	/* PHP_FFMPEG_H */


/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * indent-tabs-mode: t
 * End:
 */


ffmpeg_frame.c
/*
   This file is part of ffmpeg-php

   Copyright (C) 2004-2008 Todd Kirby (ffmpeg.php AT gmail.com)

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

   In addition, as a special exception, the copyright holders of ffmpeg-php
   give you permission to combine ffmpeg-php with code included in the
   standard release of PHP under the PHP license (or modified versions of
   such code, with unchanged license). You may copy and distribute such a
   system following the terms of the GNU GPL for ffmpeg-php and the licenses
   of the other code concerned, provided that you include the source code of
   that other code when and as the GNU GPL requires distribution of source code.

   You must obey the GNU General Public License in all respects for all of the
   code used other than standard release of PHP. If you modify this file, you
   may extend this exception to your version of the file, but you are not
   obligated to do so. If you do not wish to do so, delete this exception
   statement from your version.

*/

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "php.h"
#include "php_ini.h"
#include "php_globals.h"
#include "ext/standard/info.h"

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "php_ffmpeg.h"

#include "ffmpeg_frame.h"
#include "ffmpeg_tools.h"

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

/*
   include gd header from local include dir. This is a copy of gd.h that is
   distributed with php-5.2.5. It is distributed along with ffmpeg-php to
   allow ffmpeg-php to be built without access to the php sources
   */
#if HAVE_LIBGD20
#include "gd.h"

#define FFMPEG_PHP_FETCH_IMAGE_RESOURCE(gd_img, ret) { \
    ZEND_GET_RESOURCE_TYPE_ID(le_gd, "gd"); \
    ZEND_FETCH_RESOURCE(gd_img, gdImagePtr, ret, -1, "Image", le_gd); \
}

#if PIX_FMT_RGBA32
#define FFMPEG_PHP_FFMPEG_RGB_PIX_FORMAT PIX_FMT_RGBA32
#else
#define FFMPEG_PHP_FFMPEG_RGB_PIX_FORMAT PIX_FMT_RGB32
#endif

#define gdImageBoundsSafeMacro(im, x, y) (!((((y) < (im)->cy1) || ((y) > (im)->cy2)) || (((x) < (im)->cx1) || ((x) > (im)->cx2))))

static int le_gd; // this is only valid after calling
// FFMPEG_PHP_FETCH_IMAGE_RESOURCE()

#endif // HAVE_LIBGD20

int le_ffmpeg_frame; // not static since it is used in ffmpeg_output_movie

static zend_class_entry *ffmpeg_frame_class_entry_ptr;
zend_class_entry ffmpeg_frame_class_entry;

/* {{{ ffmpeg_frame methods[]
   Methods of the ffmpeg_frame class
   */
zend_function_entry ffmpeg_frame_class_methods[] = {

    /* object can't be created from user space so no PHP constructor */

#if HAVE_LIBGD20
    /* gd methods */
    FFMPEG_PHP_MALIAS(ffmpeg_frame, togdimage,      toGDImage,     NULL, 0)
    PHP_ME(ffmpeg_frame, __construct, NULL, 0)
#endif // HAVE_LIBGD20

    /* methods */
    FFMPEG_PHP_MALIAS(ffmpeg_frame, getwidth,                    getWidth,                   NULL, 0)
    FFMPEG_PHP_MALIAS(ffmpeg_frame, getheight,                   getHeight,                  NULL, 0)
    FFMPEG_PHP_MALIAS(ffmpeg_frame, resize,                      resize,                     NULL, 0)
    FFMPEG_PHP_MALIAS(ffmpeg_frame, iskeyframe,                  isKeyFrame,                 NULL, 0)
    FFMPEG_PHP_MALIAS(ffmpeg_frame, getpresentationtimestamp,    getPresentationTimestamp,   NULL, 0)
    FFMPEG_PHP_MALIAS(ffmpeg_frame, getpts,                      getPresentationTimestamp,   NULL, 0)
    FFMPEG_PHP_END_METHODS
};
/* }}} */


/* {{{ _php_alloc_ff_frame()
*/
static ff_frame_context* _php_alloc_ff_frame()
{
    ff_frame_context *ff_frame = NULL;

    ff_frame = emalloc(sizeof(ff_frame_context));

    ff_frame->av_frame = NULL;
    ff_frame->width = 0;
    ff_frame->height = 0;
    ff_frame->pixel_format = 0;

    return ff_frame;
}
/* }}} */


/* {{{ proto object _php_create_ffmpeg_frame()
   creates an ffmpeg_frame object, adds a ffmpeg_frame resource to the
   object, registers the resource and returns a direct pointer to the
   resource.
   */
ff_frame_context* _php_create_ffmpeg_frame(INTERNAL_FUNCTION_PARAMETERS)
{
    int ret;
    ff_frame_context *ff_frame;

    ff_frame = _php_alloc_ff_frame();

    ret = ZEND_REGISTER_RESOURCE(NULL, ff_frame, le_ffmpeg_frame);

    object_init_ex(return_value, ffmpeg_frame_class_entry_ptr);
    add_property_resource(return_value, "ffmpeg_frame", ret);
    return ff_frame;
}
/* }}} */


/* {{{ _php_free_av_frame()
*/
static void _php_free_av_frame(AVFrame *av_frame)
{
    if (av_frame) {
        if (av_frame->data[0]) {
            av_free(av_frame->data[0]);
            av_frame->data[0] = NULL;
        }
        av_free(av_frame);
    }
}
/* }}} */


/* {{{ _php_free_ffmpeg_frame()
*/
static void _php_free_ffmpeg_frame(zend_rsrc_list_entry *rsrc TSRMLS_DC)
{
    ff_frame_context *ff_frame = (ff_frame_context*)rsrc->ptr;
    _php_free_av_frame(ff_frame->av_frame);
    efree(ff_frame);
}
/* }}} */


/* {{{ register_ffmpeg_frame_class()
*/
void register_ffmpeg_frame_class(int module_number)
{
    TSRMLS_FETCH();

    le_ffmpeg_frame = zend_register_list_destructors_ex(_php_free_ffmpeg_frame,
            NULL, "ffmpeg_frame", module_number);

    INIT_CLASS_ENTRY(ffmpeg_frame_class_entry, "ffmpeg_frame",
            ffmpeg_frame_class_methods);

    /* register ffmpeg frame class */
    ffmpeg_frame_class_entry_ptr =
        zend_register_internal_class(&ffmpeg_frame_class_entry TSRMLS_CC);
}
/* }}} */


/* {{{ _php_convert_frame()
*/
int _php_convert_frame(ff_frame_context *ff_frame, int dst_fmt) {
   TSRMLS_FETCH();
   
    AVFrame *dst_frame;
    int result = 0;

    if (!ff_frame->av_frame) {
        return -1;
    }

    if (dst_fmt == ff_frame->pixel_format) {
        return 0; // NOP
    }

    dst_frame = avcodec_alloc_frame();
    av_image_alloc((AVPicture*)dst_frame, dst_fmt, ff_frame->width,
            ff_frame->height);

    result = img_convert( (AVPicture *)dst_frame, dst_fmt,
            (AVPicture *) ff_frame->av_frame, ff_frame->pixel_format, ff_frame->width, ff_frame->height);

    if (result) {
        php_error_docref(NULL TSRMLS_CC, E_WARNING, "Error converting frame");
        _php_free_av_frame(dst_frame);
    } else {
        ff_frame->av_frame = dst_frame;
        ff_frame->pixel_format = dst_fmt;
    }

    return result;
}
/* }}} */

#if HAVE_LIBGD20

/* {{{ _php_get_gd_image()
*/
static int _php_get_gd_image(int w, int h TSRMLS_DC)
{
    zval *function_name, *width, *height;
    zval **argv[2];
    zend_function *func;
    zval *retval;
    char *function_cname = "imagecreatetruecolor";
    int ret;

    if (zend_hash_find(EG(function_table), function_cname, strlen(function_cname) + 1, (void **)&func) == FAILURE) {
        php_error_docref(NULL TSRMLS_CC, E_WARNING, "Error can't find %s function", function_cname);
        return -1;
    }

    MAKE_STD_ZVAL(function_name);
    MAKE_STD_ZVAL(width);
    MAKE_STD_ZVAL(height);

    ZVAL_STRING(function_name, function_cname, 0);
    ZVAL_LONG(width, w);
    ZVAL_LONG(height, h);

    argv[0] = &width;
    argv[1] = &height;

    if (call_user_function_ex(EG(function_table), NULL, function_name,
                &retval, 2, argv, 0, NULL TSRMLS_CC) == FAILURE) {
        php_error_docref(NULL TSRMLS_CC, E_WARNING, "Error calling %s function", function_cname);

        FREE_ZVAL(function_name);
        FREE_ZVAL(width);
        FREE_ZVAL(height);
        return -1;
    }

    FREE_ZVAL(function_name);
    FREE_ZVAL(width);
    FREE_ZVAL(height);

    if (!retval || Z_TYPE_P(retval) != IS_RESOURCE) {
        php_error_docref(NULL TSRMLS_CC, E_WARNING, "Error creating GD Image");
        return -1;
    }

    ret = Z_LVAL_P(retval);
    zend_list_addref(ret);
    zval_ptr_dtor(&retval);

    return ret;
}
/* }}} */


/* {{{ _php_avframe_to_gd_image()
*/
static int _php_avframe_to_gd_image(AVFrame *frame, gdImage *dest, int width,
        int height)
{
    int x, y;
    int *src = (int*)frame->data[0];

    if (width > dest->sx || height > dest->sy) {
        return -1;
    }

    for (y = 0; y < height; y++) {
        for (x = 0; x < width; x++) {
            /* copy pixel to gdimage buffer zeroing the alpha channel */
            dest->tpixels[y][x] = src[x] & 0x00ffffff;
        }
        src += width;
    }
    return 0;
}
/* }}} */


/* {{{ _php_gd_image_to_avframe()
*/
static int _php_gd_image_to_avframe(gdImage *src, AVFrame *frame, int width,
        int height)
{
    int x, y;
    int *dest = (int*)frame->data[0];

    for (y = 0; y < height; y++) {
        for (x = 0; x < width; x++) {
            dest[x] = src->tpixels[y][x];
        }
        dest += width;
    }
    return 0;
}
/* }}} */


/* {{{ proto resource toGDImage()
*/
FFMPEG_PHP_METHOD(ffmpeg_frame, toGDImage)
{
    ff_frame_context *ff_frame;
    gdImage *gd_img;
    int rsrc_id;

    GET_FRAME_RESOURCE(getThis(), ff_frame);

    _php_convert_frame(ff_frame, FFMPEG_PHP_FFMPEG_RGB_PIX_FORMAT);

    rsrc_id = _php_get_gd_image(ff_frame->width, ff_frame->height TSRMLS_CC);

    if (rsrc_id < 0) {
        RETURN_FALSE;
    }

    RETVAL_RESOURCE(rsrc_id);

    FFMPEG_PHP_FETCH_IMAGE_RESOURCE(gd_img, &return_value);

    if (_php_avframe_to_gd_image(ff_frame->av_frame, gd_img, ff_frame->width, ff_frame->height)) {
        php_error_docref(NULL TSRMLS_CC, E_WARNING, "failed to convert frame to gd image");
        zend_list_delete(rsrc_id);
        RETURN_FALSE;
    }
    /* return the resource */
}
/* }}} */

/* {{{ proto object ffmpeg_frame(mixed)
*/
FFMPEG_PHP_METHOD(ffmpeg_frame, __construct)
{
    AVFrame *frame;
    gdImage *gd_img;
    ff_frame_context *ff_frame;
    int width, height, ret;
    zval *param;

    if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "z", &param) != SUCCESS) {
        return;
    }

    switch (Z_TYPE_P(param)) {
        case IS_STRING:
            php_error_docref(NULL TSRMLS_CC, E_WARNING, "Creating an ffmpeg_frame from a file is not implemented");
            RETURN_FALSE;
            //_php_read_frame_from_file(ff_frame, Z_STRVAL_PP(argv[0]));
            break;
        case IS_RESOURCE:
            FFMPEG_PHP_FETCH_IMAGE_RESOURCE(gd_img, &param);

            if (!gd_img->trueColor) {
                php_error_docref(NULL TSRMLS_CC, E_WARNING, "First parameter must be a truecolor GD image");
                RETURN_FALSE;
            }

            ff_frame = _php_alloc_ff_frame();

            ret = ZEND_REGISTER_RESOURCE(NULL, ff_frame, le_ffmpeg_frame);

            object_init_ex(getThis(), ffmpeg_frame_class_entry_ptr);
            add_property_resource(getThis(), "ffmpeg_frame", ret);

            width = gdImageSX(gd_img);
            height = gdImageSY(gd_img);

            /* create a an av_frame and allocate space for it */
            frame = avcodec_alloc_frame();
            avpicture_alloc((AVPicture*)frame, FFMPEG_PHP_FFMPEG_RGB_PIX_FORMAT, width, height);

            /* copy the gd image to the av_frame */
            _php_gd_image_to_avframe(gd_img, frame, width, height);

            /* set the ffmepg_frame to point to this av_frame */
            ff_frame->av_frame = frame;

            /* set the ffpmeg_frame's properties */
            ff_frame->width = width;
            ff_frame->height = height;
            ff_frame->pixel_format = FFMPEG_PHP_FFMPEG_RGB_PIX_FORMAT;
            break;
        default:
            php_error_docref(NULL TSRMLS_CC, E_WARNING, "Invalid argument passed");
            RETURN_FALSE;
    }
}
/* }}} */

#endif /* HAVE_LIBGD20 */


/* {{{ proto int getPresentationTimestamp()
*/
FFMPEG_PHP_METHOD(ffmpeg_frame, getPresentationTimestamp)
{
    ff_frame_context *ff_frame;

    GET_FRAME_RESOURCE(getThis(), ff_frame);

    RETURN_DOUBLE((double)ff_frame->pts / AV_TIME_BASE);
}
/* }}} */


/* {{{ proto int isKeyFrame()
*/
FFMPEG_PHP_METHOD(ffmpeg_frame, isKeyFrame)
{
    ff_frame_context *ff_frame;

    GET_FRAME_RESOURCE(getThis(), ff_frame);

    RETURN_LONG(ff_frame->keyframe);
}
/* }}} */


/* {{{ proto int getWidth()
*/
FFMPEG_PHP_METHOD(ffmpeg_frame, getWidth)
{
    ff_frame_context *ff_frame;

    GET_FRAME_RESOURCE(getThis(), ff_frame);

    RETURN_LONG(ff_frame->width);
}
/* }}} */


/* {{{ proto int getHeight()
*/
FFMPEG_PHP_METHOD(ffmpeg_frame, getHeight)
{
    ff_frame_context *ff_frame;

    GET_FRAME_RESOURCE(getThis(), ff_frame);

    RETURN_LONG(ff_frame->height);
}
/* }}} */


/* {{{ _php_resample_frame()
*/
int _php_resample_frame(ff_frame_context *ff_frame,
        int wanted_width, int wanted_height, int crop_top, int crop_bottom,
        int crop_left, int crop_right)
{
    AVFrame *resampled_frame;
    ImgReSampleContext *img_resample_ctx = NULL;

    if (!ff_frame->av_frame) {
        return -1;
    }

    /*
     * do nothing if width and height are the same as the frame and no
     * cropping was specified
     * */
    if (wanted_width == ff_frame->width &&
            wanted_height == ff_frame->height &&
            (!crop_left && !crop_right && !crop_top && !crop_bottom)) {
        return 0;
    }

    /* convert to AV_PIX_FMT_YUV420P required for resampling */
    _php_convert_frame(ff_frame, AV_PIX_FMT_YUV420P);

    img_resample_ctx = img_resample_full_init(
            wanted_width, wanted_height,
            ff_frame->width, ff_frame->height,
            crop_top, crop_bottom, crop_left, crop_right,
            0, 0, 0, 0);
    if (!img_resample_ctx) {
        return -1;
    }

    resampled_frame = avcodec_alloc_frame();
    av_image_alloc((AVPicture*)resampled_frame, AV_PIX_FMT_YUV420P,
            wanted_width, wanted_height);

    img_resample(img_resample_ctx, (AVPicture*)resampled_frame,
            (AVPicture*)ff_frame->av_frame);

    _php_free_av_frame(ff_frame->av_frame);

    img_resample_close(img_resample_ctx);

    ff_frame->av_frame = resampled_frame;
    ff_frame->width = wanted_width;
    ff_frame->height = wanted_height;

    return 0;
}
/* }}} */


/* {{{ proto boolean resize(int width, int height [, int crop_top [, int crop_bottom [, int crop_left [, int crop_right ]]]])
*/
FFMPEG_PHP_METHOD(ffmpeg_frame, resize)
{
    ff_frame_context *ff_frame = NULL;
    long wanted_width, wanted_height;
    long crop_top = 0, crop_bottom = 0, crop_left = 0, crop_right = 0;
    int res;

    GET_FRAME_RESOURCE(getThis(), ff_frame);

    if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "ll|llll", &wanted_width, &wanted_height, &crop_top, &crop_bottom, &crop_left, &crop_right) != SUCCESS) {
        return;
    }

    if (crop_right % 2) {
        php_error_docref(NULL TSRMLS_CC, E_WARNING, "Crop right must be an even number");
        RETURN_FALSE;
    }

    if (crop_left % 2) {
        php_error_docref(NULL TSRMLS_CC, E_WARNING, "Crop left must be an even number");
        RETURN_FALSE;
    }

    if (crop_bottom % 2) {
        php_error_docref(NULL TSRMLS_CC, E_WARNING, "Crop bottom must be an even number");
        RETURN_FALSE;
    }

    if (crop_top % 2) {
        php_error_docref(NULL TSRMLS_CC, E_WARNING, "Crop top be an even number");
        RETURN_FALSE;
    }

    /* bounds check wanted height */
    if (wanted_height < 1) {
        php_error_docref(NULL TSRMLS_CC, E_WARNING, "Frame height must be greater than zero");
        RETURN_FALSE;
    }

    /* wanted height must be even number for lavc resample */
    if (wanted_height % 2) {
        php_error_docref(NULL TSRMLS_CC, E_WARNING, "Frame height must be an even number");
        RETURN_FALSE;
    }

    /* bounds check wanted width */
    if (wanted_width < 1) {
        php_error_docref(NULL TSRMLS_CC, E_WARNING, "Frame width must be greater than zero");
        RETURN_FALSE;
    }

    /* wanted width must be even number for lavc resample */
    if (wanted_width % 2) {
        php_error_docref(NULL TSRMLS_CC, E_WARNING, "Frame width must be an even number");
        RETURN_FALSE;
    }

    /* resize frame */
    res = _php_resample_frame(ff_frame, wanted_width, wanted_height,
            crop_top, crop_bottom, crop_left, crop_right);

    if (res < 0) {
        RETURN_FALSE;
    }

    RETURN_TRUE;
}
/* }}} */


/* {{{ proto boolean crop([, int crop_top [, int crop_bottom [, int crop_left [, int crop_right ]]]])
*/
PHP_FUNCTION(crop)
{
    ff_frame_context *ff_frame;
    long crop_top = 0, crop_bottom = 0, crop_left = 0, crop_right = 0;
    int res;

    GET_FRAME_RESOURCE(getThis(), ff_frame);

    if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "|llll", &crop_top, &crop_bottom, &crop_left, &crop_bottom) != SUCCESS) {
        return;
    }

    if (crop_right % 2) {
        php_error_docref(NULL TSRMLS_CC, E_WARNING, "Crop right must be an even number");
        RETURN_FALSE;
    }

    if (crop_left % 2) {
        php_error_docref(NULL TSRMLS_CC, E_WARNING, "Crop left must be an even number");
        RETURN_FALSE;
    }

    if (crop_bottom % 2) {
        php_error_docref(NULL TSRMLS_CC, E_WARNING, "Crop bottom must be an even number");
        RETURN_FALSE;
    }

    if (crop_top % 2) {
        php_error_docref(NULL TSRMLS_CC, E_WARNING, "Crop top be an even number");
        RETURN_FALSE;
    }

    /* resample with same dimensions */
    res = _php_resample_frame(ff_frame, ff_frame->width, ff_frame->height,
            crop_top, crop_bottom, crop_left, crop_right);

    if (res < 0) {
        RETURN_FALSE;
    }

    RETURN_TRUE;
}
/* }}} */


/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4 expandtab
 * vim<600: noet sw=4 ts=4
 */
